nixos: Support fileSystems.<name>.depends with fstab-generator

ElvishJerricco opened the pull request:
###### Description of changes

I have barely tested this, so I'm not entirely sure it's correct. Closes #217179

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ElvishJerricco committed:
nixos: Support fileSystems.<name>.depends with fstab-generator

RaitoBezarius commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -1098,20 +1098,23 @@ in
           noCheck = true; # fsck fails on a r/o filesystem
         };
       }
+      (lib.mkIf (config.boot.initrd.systemd.enable && cfg.writableStore) {
```
FYI, in the diskless VM PR, I will introduce this definitely and remove any manual mount.

astro replied:
What is the diskless VM PR?

ElvishJerricco commented:
This would fix #217179

ElvishJerricco committed:
overlayfs: Use depends

ElvishJerricco committed:
overlayfs: Assert that upperdir == null requires systemd initrd

ElvishJerricco commented:
@nikstur I updated this with a little bit of tweaking to the overlay fs stuff. The `depends` functionality should be the same on systemd initrd, with *slightly* better behavior in scripted initrd. While testing with scripted, I noticed it doesn't work with the ro overlays, so I just added an assertion to disallow that, rather than trying to implement it.

philiptaron commented on nixos/modules/tasks/filesystems/overlayfs.nix:
```diff
@@ -123,18 +127,29 @@ in
+          ]) overlayFileSystems
+        )
+        ++ lib.mapAttrsToList (_: fs: {
+          assertion = fs.upperdir == null -> config.boot.initrd.systemd.enable;
```
_This_ is nice and clear as an implication.

philiptaron commented on nixos/tests/filesystems-overlayfs.nix:
```diff
@@ -26,7 +26,6 @@ in
 
   nodes.machine = { config, pkgs, ... }: {
     boot.initrd.systemd.enable = true;
-    boot.initrd.availableKernelModules = [ "overlay" ];
```
Help me understand why this needed to leave? Is it just excess to requirements now?

ElvishJerricco replied:
It's just redundant is all. It never needed to be there in the first place.

philiptaron commented:
I'll note that ofBorg is unhappy for the same reason that I created https://github.com/NixOS/nixpkgs/pull/320846: the value of the `message` depends on the assertion failing, and as written it doesn't protect against that being so.

ElvishJerricco commented on nixos/modules/tasks/filesystems.nix:
```diff
@@ -140,6 +140,15 @@ let
       (mkIf (utils.fsNeededForBoot config) [ "x-initrd.mount" ])
+      (mkIf
+        # With scripted stage 1, depends is implemented by sorting 'config.system.build.fileSystems'
+        (config.depends != [ ] && (utils.fsNeededForBoot config -> _config.boot.initrd.systemd.enable))
```
> and in every case except for _config.boot.initrd.systemd.enable, add these x-systemd.requires-mounts-for lines.

No, the only time we *don't* add the `x-systemd.requires-mounts-for=` option is when it's `fsNeededForBoot` and `boot.initrd.systemd.enable == false`, which is what your truth table shows. i.e. If the file system is mounted in scripted initrd, we let scripted initrd's semantics handle `depends` (which, as the comment indicates, is handled by the sorting of file systems).

> It's not clear immediately what the semantics of x-systemd.requires-mounts-for= (with nothing after it) is. That's the case when utils.fsNeededForBoot config is false

No it isn't. Re-parenthesizing, it's `"x-systemd.requires-mounts-for=${(optionalString (utils.fsNeededForBoot config) "/sysroot") + x}"`, i.e. the `x` string is unconditional and always included. There should never be a `x-systemd.requires-mounts-for=` with no value.

ElvishJerricco replied:
> This is the first and only time that /sysroot is named in this file. Is there a place that could be linked to document or reference the structure of the initrd?

The only real documentation that talks about `/sysroot` is `man bootup`, but we reference `/sysroot` in several places for systemd initrd so I'm not sure why this particular instance needs a link to documentation when the others don't.

philiptaron replied:
> what your truth table shows

Yeah, I thought I had typed `== false` but obviously not. We're in agreement here about what's happening.

> Re-parenthesizing

That change makes this **significantly** clearer. I did not understand how this would render, as my commentary shows.

philiptaron commented on nixos/modules/tasks/filesystems.nix:
```diff
@@ -136,10 +136,21 @@ let
+        (lib.length config.depends > 0 && (inInitrd -> moduleArgs.config.boot.initrd.systemd.enable))
+        (
+          map (
+            x: "x-systemd.requires-mounts-for=${optionalString inInitrd "/sysroot" + x}"
```
What about:

```suggestion
            x: "x-systemd.requires-mounts-for=${optionalString inInitrd "/sysroot"}${x}"
```

ElvishJerricco replied:
Done

philiptaron reviewed (commented):
This looks great. I'm going to give it a whirl on my home machine and report back.

philiptaron reviewed (approved):
Tested and works.
